python/mxnet/gluon/contrib/parallel.py

```diff
@@ -0,0 +1,362 @@
+
+    def _clear(self):
+        self.list = []
+        self.push_tasks = self.counter
+        self.reduce_tasks = self.counter
+
+    def __len__(self):
+        return len(self.list)
+
+    def __repr__(self):
+        return 'Barrier'
+
+
+class DataParallelModel(object):
+    """Data Parallelism
+
+    Hide the difference of single/multiple GPUs to the user.
+    This container parallelizes the application of the given module by
+    splitting the input across the specified devices by chunking in the
+    batch dimension.
+    In the forward pass, the module is replicated on each device,
+    and each replica handles a portion of the input. During the backwards pass,
+    gradients from each replica are summed into the original module.
+    Note that the outputs are not gathered, please use compatible
+    :class:`mxnet.gluon.contrib.DataParallelCriterion`.
+
+    The batch size should be larger than the number of GPUs used. It should
+    also be an integer multiple of the number of GPUs so that each chunk is
+    the same size (so that each GPU processes the same number of samples).
+
+    Parameters
+    ----------
+    module : object
+        Network to be parallelized.
+    ctx_list : list
+        A list of contexts
+    sync : bool
+        enable synchronization (default: False).
+
+
+    Inputs:
+        - **inputs**: list of input (NDArrays)
+
+    Outputs:
+        - **outputs**: list of output (NDArrays)
+
+    Example::
+        >>> ctx = [mx.gpu(0), mx.gpu(1)]
+        >>> net = DataParallelModel(model, ctx_list=ctx)
+        >>> y = net(x)
+    """
+    def __init__(self, module, ctx_list=None, sync=False):
+        module.collect_params().reset_ctx(ctx=ctx_list)
+        self.ctx_list = ctx_list
+        self.module = module
+        self.sync = sync
+
+    def __call__(self, *inputs, **kwargs):
+        if not self.ctx_list:
+            return self.module(*inputs, **kwargs)
+        inputs, kwargs = _split_load_kwargs(inputs, kwargs, self.ctx_list)
+        assert(len(inputs) == len(self.ctx_list))
+        if len(self.ctx_list) == 1:
+            return tuple([tuple_map(self.module(*inputs[0], **kwargs[0]))])
+        return parallel_apply(self.module, inputs, kwargs, self.sync)
+
+    def __repr__(self):
+        return 'DataParallel:\n module = {' + self.module.__repr__() + '}'
+
+
+class DataParallelCriterion(object):
+    """Data Parallelism
+
+    Hide the difference of single/multiple GPUs to the user.
+    The targets are splitted across the specified devices by chunking in
+    the batch dimension.
+    In the forward pass, the module is replicated on each device,
+    and each replica handles a portion of the input. During the backwards pass,
+    gradients from each replica are summed into the original module.
+    Note that the outputs are not gathered, please use compatible
+    :class:`mxnet.gluon.contrib.DataParallelModel`
+
+    Parameters
+    ----------
+    module : object
```

Comment:
Is it better to call it block instead of module?

Reply:
Block has a general meaning in Gluon. We can probably call it model, if you think that is better.